### PR TITLE
fixes #6248 - API V2 return object for POST/PUT/DELETE should *not* include root node

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by domcleal** [comment]
Won't this change v1 too?

**Comment by isratrade** [comment]
@daviddavis, ping. Can you check if this fix makes this more consistent. The object returned on POST/PUT/DELETE now has no node. I don't think it affects anything on json parameters received or GET request that use RABL templates.

**Comment by isratrade** [comment]
@domcleal, yes, it will change v1. I guess it's a no go.

**Comment by isratrade** [comment]
@daviddavis, this will probably not be merged per @domcleal's comment above.

**Comment by isratrade** [comment]
@daviddavis, there are also several test failures locally caused by this change.

**Comment by daviddavis** [comment]
So, to sum up, I guess our options are:
1. Leave it as is for now (hopefully we can find a fix in the future)
2. Change katello to return root nodes for POST/PUT/DELETE calls to be more consistent
3. Find another solution that only affects v2.

I'm fine with #1 suppose. I think #2 is a bad idea. #3 seems best purely in theory but it's not feasible? 

**EDIT**: Would there be a way to use the show rabl templates for DELETE/PUT/POST requests? I believe this is what we're doing in Katello. This sounds like the best in terms of consistency not only with Katello but also with show requests within foreman.

**Comment by waldenraines** [comment]
Is the third option really not feasible?  I'm fine with option 1 or option 3.  The second option would be largely destabilizing to katello.

**Comment by thomasmckay** [comment]
we agreed a long time ago to the non-nested format
http://theforeman.org/manuals/1.5/index.html#5.1.3JSONResponseFormatforSingleObjects

File: app/controllers/api/v2/base_controller.rb [hunk 1]
```diff
@@ -9,6 +9,11 @@ class BaseController < Api::BaseController
       end
 
       before_filter :setup_has_many_params, :only => [:create, :update]
+      # override true value defined in config/initializers/wrap_parameters.rb for V2 responses for POST, PUT, DELETE
+      before_filter :except => [:index] do
+        resource_class.include_root_in_json = false
```
Comment (isratrade) [inline app/controllers/api/v2/base_controller.rb]: @waldenraines, @daviddavis, I reviewed katello code and see how you are rendering RABL templates for each each action with `respond_for_#{action}`

Foreman is essentially just doing `render :json => @resource` for the response for PUT/POST/DELETE

This fix changing `resource_class.include_root_in_json = false` in Api::V2::BaseController fixes the issue without affecting V1.

In doing so, I noticed the inflection bug with "HostClass" and "HostgroupClass"

If this `before_filter` is also on the `index` actions, then several tests fail (dashboard, statistics, autosign, etc). I didn't investigate these failures as they don't affect our consistency in documentation between katello and foreman.  All tests pass with `:except => [:index]`

**Comment by isratrade** [comment]
[test]

**Comment by daviddavis** [comment]
@isratrade thanks for fixing this.

**Comment by isratrade** [comment]
[test] interesting that `SettingsControllerTest` fails. This patch shouldn't touch it. My assumption is that once `Setting.include_root_in_json = false` is run by the API v2 controller, it persists for the model and affects everything else including API v1, so this solution won't work. Will need to re-think a new solution.

**Comment by daviddavis** [comment]
@isratrade interesting. what about putting the before filter into api base controller and then set include_root_in_json to true or false based on whether it's v1 or v2 respectively.

**Comment by daviddavis** [comment]
Also, since we're close to the end of the release and because this may cause instability, do we maybe want to wait on merging it?

**Comment by isratrade** [comment]
@domcleal, @daviddavis, this should work for V1 and V2 as expected. All tests pass.

**Comment by daviddavis** [comment]
@isratrade can you update your commit to point to http://projects.theforeman.org/issues/6248? #5880 got accidentally deleted. Thanks.

File: test/functional/settings_controller_test.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ def test_update_valid
     assert :success
     response = ActiveSupport::JSON.decode(@response.body)
     assert !response.empty?
-    assert_equal new_value, response['setting']['value']
+    assert_equal new_value, response['value']
```
Comment (domcleal) [inline test/functional/settings_controller_test.rb]: This indicates a break in the UI which uses AJAX to update the saved value when editing settings.  Mine's showing a load of AJAX when I change a value now.

I guess the change to the default will affect all UI controllers - do we need to add a before_filter in the ApplicationController so it remains true for them?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -13,6 +13,8 @@ class ApplicationController < ActionController::Base
   helper 'layout'
   helper_method :authorizer
 
+  # ensure include_root_in_json = true
+  before_filter { ActiveRecord::Base.include_root_in_json = true }
```
Comment (isratrade) [inline app/controllers/application_controller.rb]: added this and re-committed.

File: app/controllers/api/v2/base_controller.rb [hunk 4]
```diff
@@ -9,6 +9,9 @@ class BaseController < Api::BaseController
       end
 
       before_filter :setup_has_many_params, :only => [:create, :update]
+      # ensure include_root_in_json = false in case it was overridden by Api::V1 to true
```
Comment (domcleal) [inline app/controllers/api/v2/base_controller.rb]: Perhaps use an around_filter I think, then we can restore it back to the previous value reliably?

Comment (isratrade) [inline app/controllers/api/v2/base_controller.rb]: @domcleal, that makes sense. good idea.  will do.

File: app/controllers/api/v2/base_controller.rb [hunk 5]
```diff
@@ -9,6 +9,9 @@ class BaseController < Api::BaseController
       end
 
       before_filter :setup_has_many_params, :only => [:create, :update]
+      # ensure include_root_in_json = false for V2 only
+      around_filter :disable_json_root
```
Comment (isratrade) [inline app/controllers/api/v2/base_controller.rb:13]: updated PR using `around_filter`

File: app/controllers/api/v2/base_controller.rb [hunk 6]
```diff
@@ -74,6 +77,17 @@ def render_error(error, options = { })
...
                              :layout   => 'api/v2/layouts/error_layout')
       end
+
+      private
+
+      def disable_json_root
+        # disable json root element
+        ActiveRecord::Base.include_root_in_json = false
+        yield
+        # re-enable json root element
```
Comment (domcleal) [inline app/controllers/api/v2/base_controller.rb:88]: should be within `ensure`

Comment (isratrade) [inline app/controllers/api/v2/base_controller.rb:88]: @domcleal, you're right. re-committed.

**Comment by domcleal** [comment]
Merged as 301e9dc for Foreman 1.6.0, thanks @isratrade.

**Pull request closed without merging**